Speed up `emitJson` replacement

dolio opened the pull request:
This PR includes some improvements to the `emitJson` replacement in the runtime.

- Switched to using a strict text builder instead of aggregating up `IO Text`. Exceptions are managed via a final `evaluate` instead.
- Made the function produce a single chunk rope, which speeds up eventual conversion to bytes.
- Improved the predicate looking for characters to escape, which took a considerable portion of the time in a lot of cases (anything with lots of objects has a lot of strings to check).

dolio committed:
Speed up `emitJson` replacement

- Switched to using a strict text builder instead of aggregating up
  `IO Text`. Exceptions are managed via a final `evaluate` instead.
- Made the function produce a single chunk rope, which speeds up eventual
  conversion to bytes.
- Improved the predicate looking for characters to escape, which took a
  considerable amount of the time in a lot of cases.

dolio committed:
automatically run ormolu

aryairani commented:
does this need dogfooding or is it pretty solid? i'm thinking about doing another release soon

dolio commented:
Maybe hold off until after just in case.

It seems like the one-chunk thing might make parsing slower (investigating). That's kind of a silly use case (emit then parse), but I suppose there could be other gotchas.

Edit: think I know why. The way the parser is written, it calls some text functions (like `splitAt`) that call `length` on chunks, but with one enormous chunk that's bad because `Text` is `O(n)` length.

dolio committed:
Use a custom lazy Text `splitAt` in json parsing

The one from the library calls `length` on chunks, which is very bad for
huge chunks, and seems like it might even be costly on more modest chunks.

dolio committed:
automatically run ormolu

dolio commented:
Wrote a custom version of `splitAt`. Now the single-chunk parse seems to be faster than the version running on modest sized chunks.

dolio commented:
Manually ran CI against the ormolu commit here:

https://github.com/unisonweb/unison/actions/runs/20247859959

I think it fixed the problem from the previous commit.
